shell/gpu/gpu_surface_metal_impeller.mm

```diff
@@ -118,8 +118,13 @@
           return surface->Present();
         }
 
-        impeller::DlDispatcher impeller_dispatcher;
-        display_list->Dispatch(impeller_dispatcher);
+        if (!buffer_damage.has_value()) {
+          auto size = surface->GetSize();
+          buffer_damage = SkIRect::MakeWH(size.width, size.height);
+          clip_rect = impeller::IRect::MakeXYWH(0, 0, size.width, size.height);
+        }
+        impeller::DlDispatcher impeller_dispatcher(clip_rect.value());
+        display_list->Dispatch(impeller_dispatcher, buffer_damage.value());
```

Comment:
I implemented this a bit differently for impeller, rather than adding a clip - i added a translate and then shrunk the size of the root render pass, which causes impeller to cull.

translate: https://github.com/flutter/engine/blob/main/flow/compositor_context.cc#L189

Shrink: https://github.com/flutter/engine/blob/main/impeller/renderer/backend/metal/surface_mtl.mm#L57

I'm not sure if this clip would already do the right thing though.

Reply:
> which causes impeller to cull.

Causes it to cull where?

impeller::Dispatcher::DrawRect immediately calls Canvas::DrawRect without any tests.
Canvas::DrawRect (sometimes does something different for strokes and mask blurs) immediately creates and adds an entity without any other checks.

Where is the culling done if that rectangle was so far outside the clip that it isn't visible? The Dispatch method on DL can do this kind of bounds culling and then the impeller::Dispatcher never even sees the op...

Reply:
I'm not sure this code here can apply a translation. If there is no damage rect, the cull rect is the size of the surface so it is already sized correctly and 0-based. If there is a damage rect, then the remainder of the surface must remain intact so we need to render the culled ops "in situ" with respect to origin and surface size.

While I still don't think I see all of the relevance of your comment, I don't think we can do translations and resizes of surfaces here because of the nature of this code and the damage rect...?

Reply:
Ding ding! With your later comment that the blinking cursor was missing I think I finally got what you are saying. Unlike the other platforms that reuse the original framebuffer surface to render the partial damage inside a clip in the middle of the surface, Impeller renders the damage to a reduced size surface and then blits it into place?

In that case then the Impeller cull rect should always be set to just the rect(size) value with no origin, but the Dispatch call should be called with the damage rect since it is in the coordinate space of the original frame?

Reply:
Makes sense to me

Reply:
Looking again, it looks like the translate happens before the frame DisplayList is generated, so both rects should be 0-origin-based?

And, doing all of that fixes the blinking cursor problem...